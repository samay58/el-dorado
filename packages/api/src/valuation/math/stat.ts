// Statistical helper functions (e.g., z-score) 

/**
 * Calculates the z-score of a value given the mean and standard deviation.
 * Z-score indicates how many standard deviations an element is from the mean.
 *
 * @param x The value for which to calculate the z-score.
 * @param mean The mean (μ) of the dataset.
 * @param stdDev The standard deviation (σ) of the dataset.
 * @returns The z-score, or 0 if standard deviation is 0.
 */
export function zScore(x: number, mean: number, stdDev: number): number {
  if (stdDev === 0) {
    // If standard deviation is 0, it means all values are the same as the mean.
    // So, any value x would be equal to the mean, hence z-score is 0.
    // Or, it can be seen as x is 0 standard deviations away from the mean.
    return 0;
  }
  return (x - mean) / stdDev;
} 