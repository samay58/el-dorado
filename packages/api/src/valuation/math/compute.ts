import { VALUATION_WEIGHTS } from '../config';
import { zScore } from './stat';
import { DerivedSignals } from '../signals/zillowDerived';
import { RedfinSignals } from '../signals/redfin'; // Assuming stubs might provide some optional fields

// Combine all possible signal sources into one type for computeValue
export interface AllSignals extends DerivedSignals, RedfinSignals {
  // Add any other signals that might come from different sources
  hotFlag?: number; // Example: 0 or 1, sourced from Redfin or elsewhere eventually
  // New public data signals will be added here, e.g.:
  // listPriceToMedianZipRatio?: number;
  // ppsfToMedianZipRatio?: number;
  // domToMedianZipRatio?: number;
  // zhviAppreciation12Mo?: number;
  // marketHotnessScore?: number;
  // hedonicDelta?: number; // From new hedonic model
}

export interface ValueComputationResult {
  valueIndex: number;
  rawScore: number; // Sum of weighted contributions before any final scaling if needed
  breakdown: Record<string, number>; // Weighted contribution of each signal
  // Optionally include z-scores themselves if useful for debugging
  // zScores?: Record<string, number>; 
}

// --- PLACEHOLDER STATISTICS ---
// These will be re-evaluated based on new public data signals.
const STATS = {
  deltaZ:     { mean: 0.05, stdDev: 0.15 },  
  domPct:     { mean: 1.0,  stdDev: 0.5  },
  // New stats for new signals will be added here
  // listPriceToMedianZipRatio: { mean: 1.0, stdDev: 0.2 },
  // ppsfToMedianZipRatio: { mean: 1.0, stdDev: 0.2 },
  // domToMedianZipRatio: { mean: 1.0, stdDev: 0.3 },
};
// --- END PLACEHOLDER STATISTICS ---

/**
 * Computes the valueIndex and its components from a set of signals.
 * This function will be significantly refactored based on new public data signals.
 * Old formula (for reference, will be replaced):
 *   -1.0 * z(deltaZ) 
 *   -0.6 * z(domPct) 
 *   -0.4 * recentCut 
 *   +0.5 * hotFlag 
 *   -0.3 * z(hcDelta ?? deltaZ) 
 *   -0.2 * z(attomDelta ?? deltaZ);
 */
export function computeValue(signals: AllSignals): ValueComputationResult {
  const breakdown: Record<string, number> = {};
  let rawScore = 0;

  // deltaZ (from Zillow extract)
  if (signals.deltaZ !== undefined) {
    const zDeltaZ = zScore(signals.deltaZ, STATS.deltaZ.mean, STATS.deltaZ.stdDev);
    const weightedZDeltaZ = VALUATION_WEIGHTS.deltaZ * zDeltaZ;
    breakdown.deltaZ = weightedZDeltaZ;
    rawScore += weightedZDeltaZ;
  }

  // domPct (from Zillow extract, relative to SF_MEDIAN_DOM)
  if (signals.domPct !== undefined) {
    const zDomPct = zScore(signals.domPct, STATS.domPct.mean, STATS.domPct.stdDev);
    const weightedZDomPct = VALUATION_WEIGHTS.domPct * zDomPct;
    breakdown.domPct = weightedZDomPct;
    rawScore += weightedZDomPct;
  }

  // recentCut (from Zillow extract, not z-scored, used directly)
  if (signals.recentCut !== undefined) {
    const weightedRecentCut = VALUATION_WEIGHTS.recentCut * signals.recentCut;
    breakdown.recentCut = weightedRecentCut;
    rawScore += weightedRecentCut;
  }

  // hotFlag (placeholder, will be from Realtor.com or Redfin)
  const hotFlagValue = signals.hotFlag !== undefined ? signals.hotFlag : 0; 
  const weightedHotFlag = VALUATION_WEIGHTS.hotFlag * hotFlagValue;
  breakdown.hotFlag = weightedHotFlag;
  rawScore += weightedHotFlag;

  // TODO: Implement new signal processing based on public data
  // e.g., listPriceToMedianZipRatio, zhviAppreciation12Mo, marketHotnessScore, hedonicDelta
  // Each will have its own weight, potential z-scoring, and addition to rawScore & breakdown.

  // Removed hcDelta and attomDelta logic
  
  return {
    valueIndex: parseFloat(rawScore.toFixed(2)), 
    rawScore: parseFloat(rawScore.toFixed(4)), 
    breakdown,
  };
} 