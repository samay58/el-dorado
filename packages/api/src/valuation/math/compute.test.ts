import { describe, it, expect } from 'vitest';
import { computeValue, AllSignals } from '../math/compute';
// import { VALUATION_WEIGHTS } from '../config'; // For property-based tests if needed
// import * as fc from 'fast-check'; // For property-based tests

// --- Mock/Example Fixture Data ---
// Ideally, load these from JSON files in __fixtures__
// For demonstration, defining inline. Replace with actual fixture loading.
const mockZillowExtractZpid12345 = {
  zpid: '12345',
  listPrice: 2000000,
  zestimate: 1900000,
  daysOnSite: 30,
  priceHistory: [
    { event: 'Listed for sale', date: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString(), price: 2000000 },
  ],
  // latitude, longitude, zip would come from actual ZillowExtract
};

const mockZillowExtractZpid67890 = {
  zpid: '67890',
  listPrice: 1500000,
  zestimate: 1600000,
  daysOnSite: 70,
  priceHistory: [
    { event: 'Price Change', date: new Date(Date.now() - 10 * 24 * 60 * 60 * 1000).toISOString(), price: 1500000 },
    { event: 'Listed for sale', date: new Date(Date.now() - 70 * 24 * 60 * 60 * 1000).toISOString(), price: 1550000 },
  ],
};

// --- Helper to create AllSignals from a ZillowExtract mock ---
// This would need to be more sophisticated if Redfin/HC/ATTOM signals were real
const createSignalsForTest = (zillowData: any, otherSignals?: Partial<AllSignals>): AllSignals => {
  // Simulate deriving signals from Zillow data (simplified version of zillowDerived.ts logic)
  const derived = {
    deltaZ: (zillowData.listPrice && zillowData.zestimate) ? (zillowData.listPrice - zillowData.zestimate) / zillowData.zestimate : undefined,
    domPct: zillowData.daysOnSite ? zillowData.daysOnSite / 55 : undefined, // Assuming SF_MEDIAN_DOM = 55
    recentCut: 0, // Simplified: needs proper PriceHistoryEvent[] and logic from zillowDerived.ts
  };
  if (zillowData.priceHistory && zillowData.priceHistory.length > 1 && /price change/i.test(zillowData.priceHistory[0].event) && zillowData.priceHistory[0].price < zillowData.priceHistory[1].price) {
    const eventDate = new Date(zillowData.priceHistory[0].date);
    const thirtyDaysAgo = new Date();
    thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
    if (eventDate >= thirtyDaysAgo) {
      derived.recentCut = 1;
    }
  }

  return {
    ...derived,
    ...otherSignals,
  };
};

describe('computeValue', () => {
  it('should calculate valueIndex and breakdown correctly for fixture 1 (ZPID 12345)', () => {
    const signals = createSignalsForTest(mockZillowExtractZpid12345, { hotFlag: 0 });
    const result = computeValue(signals);

    // Snapshot testing is excellent here
    expect(result.valueIndex).toMatchSnapshot('valueIndex_zpid12345');
    expect(result.rawScore).toMatchSnapshot('rawScore_zpid12345');
    expect(result.breakdown).toMatchSnapshot('breakdown_zpid12345');
    
    // Example explicit checks (less maintainable than snapshots for complex objects)
    // expect(result.valueIndex).toBeCloseTo(-0.78, 2); // Value depends heavily on placeholder STATS
  });

  it('should calculate valueIndex and breakdown correctly for fixture 2 (ZPID 67890)', () => {
    const signals = createSignalsForTest(mockZillowExtractZpid67890, { hotFlag: 1, hcDelta: -0.05 });
    const result = computeValue(signals);

    expect(result.valueIndex).toMatchSnapshot('valueIndex_zpid67890');
    expect(result.rawScore).toMatchSnapshot('rawScore_zpid67890');
    expect(result.breakdown).toMatchSnapshot('breakdown_zpid67890');
  });

  it('should handle missing optional signals gracefully', () => {
    const signals: AllSignals = {
      // Only provide mandatory ones, or even fewer if computeValue can handle it
      deltaZ: 0.1,
      domPct: 1.2,
      // recentCut, hotFlag, hcDelta, attomDelta are missing
    };
    const result = computeValue(signals);
    // Check that it doesn't crash and produces a number
    expect(typeof result.valueIndex).toBe('number');
    expect(result.breakdown.recentCut).toBeUndefined(); // Or expect it to be 0 if that's the default
    expect(result.breakdown.hotFlag).toBeDefined(); // hotFlag defaults to 0 * weight
    // Add more assertions based on expected behavior with missing signals
  });

  // --- Property-Based Tests (using fast-check) ---
  // describe('property-based tests', () => {
  //   it('valueIndex should decrease as deltaZ increases (ceteris paribus)', () => {
  //     fc.assert(
  //       fc.property(fc.double({ min: -0.5, max: 0.5, noNaN: true }), deltaZ => {
  //         const signals1: AllSignals = { deltaZ: deltaZ, domPct: 1.0, recentCut: 0, hotFlag: 0 };
  //         const signals2: AllSignals = { deltaZ: deltaZ + 0.1, domPct: 1.0, recentCut: 0, hotFlag: 0 };
  //         const result1 = computeValue(signals1);
  //         const result2 = computeValue(signals2);
  //         // VALUATION_WEIGHTS.deltaZ is negative (-1.0)
  //         expect(result2.valueIndex).toBeLessThanOrEqual(result1.valueIndex);
  //       })
  //     );
  //   });

  //   // Add more property-based tests for other signals and their expected impact on valueIndex
  //   // e.g., valueIndex should decrease as domPct increases (weight is -0.6)
  //   // e.g., valueIndex should decrease if recentCut is 1 (weight is -0.4)
  //   // e.g., valueIndex should increase if hotFlag is 1 (weight is +0.5)
  // });
}); 