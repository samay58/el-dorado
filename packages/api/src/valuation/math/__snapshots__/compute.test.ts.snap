// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 1 (ZPID 12345) > breakdown_zpid12345 1`] = `
{
  "deltaZ": -0.01754385964912277,
  "domPct": 0.5454545454545455,
  "hotFlag": 0,
  "recentCut": -0,
}
`;

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 1 (ZPID 12345) > rawScore_zpid12345 1`] = `0.5279`;

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 1 (ZPID 12345) > valueIndex_zpid12345 1`] = `0.53`;

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 2 (ZPID 67890) > breakdown_zpid67890 1`] = `
{
  "deltaZ": 0.75,
  "domPct": -0.3272727272727272,
  "hotFlag": 0.5,
  "recentCut": -0.4,
}
`;

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 2 (ZPID 67890) > rawScore_zpid67890 1`] = `0.5227`;

exports[`computeValue > should calculate valueIndex and breakdown correctly for fixture 2 (ZPID 67890) > valueIndex_zpid67890 1`] = `0.52`;
